### PR TITLE
CSS cloned site token

### PR DESCRIPTION
**Pull request by ranok**
## Proposed changes

This PR adds a new token type that uses CSS to detect a cloned (phishing) site by using a unique image URL that checks the Referer header in a CloudFront Function and if the Referer doesn't match the expected value, it redirects to the token server for alerting.

## Types of changes

What types of changes does your code introduce to this repository?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] Lint and unit tests pass locally with my changes (if applicable)
- [ ] I have run pre-commit (`pre-commit` in the repo)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Linked to the relevant github issue or github discussion

## Further comments

By default, this generates a CSS file for use with Azure AD Company Branding, but the url() can be put into any HTML element with the expected referer field set. microsoftonline.com is the expected Azure referer.

### CODE REVIEW

**Comment by ranok** [comment]
These next commits allow for automatic deployment into a tenant. There's probably some UX love that's needed, and more robust error handling, but it's a good baseline.

**Comment by ranok** [comment]
At this point, I think there are a few things needed:

- [ ] Make the installer Entra application come from the official Azure tenant, and ensure it's using a verified domain
- [ ] If the one-click installer finds an existing installed CSS, it should error more gracefully to the front-end, and provide the user the ability to download the CSS snippet to manually add and upload
- [ ] If the installer has any other issues, it should send them to the front-end
- [ ] Once installed, the installer should verify installation, and pass that information to the front-end instead of immediately closing the authorization pop-up.

File: canarytokens/tokens.py [hunk 1]
```diff
@@ -454,6 +454,27 @@ def _get_response_for_clonedsite(
         request.setHeader("Content-Type", "image/gif")
         return GIF
 
+    @staticmethod
+    def _get_info_for_cssclonedsite(request: Request):
+        http_general_info = Canarytoken._grab_http_general_info(request=request)
+
+        referer = request.getHeader('Referer')
+        r_arg = request.args.get(b"r", [None])[0]
+        if r_arg != None:
```
Comment (jayjb) [inline canarytokens/tokens.py:463]: Can we change this to:
```
if r_arg is not None:
```

File: frontend/app.py [hunk 2]
```diff
@@ -530,6 +537,23 @@ async def download(
...
+    """
+    This page is loaded after a user has authN and authZ'd into their tenant and granted the permissions to install the CSS
+    Once the CSS is installed into their tenant, and we revoke our permission grants, we can close the window as this will happen in 
+    a pop-up context.
+    """
+    if admin_consent == 'True':
+        tenant_id = tenant
+        if css := state:
+            css = b64decode(unquote(state)).decode()
+        if css != None and tenant_id != None:
```
Comment (jayjb) [inline frontend/app.py:554]:
```
if css is not None and tenant_id is not None:
```

**Review by jayjb — CHANGES_REQUESTED** [review]

**Comment by ranok** [comment]
I set up a framework for passing information from the Azure auto-installer to the front-end in 69f707d. Will need styling now that it doesn't immediately close itself.

File: aws-css-token-infra/CSSClonedSiteCFFunc/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// Cloudfront Function to check the referrer compared to the expected one base64 encoded into the URI
+// Expected uri looks like: /TOKEN_ID/escape(btoa(expected_referrer))/imagename.gif
+// Either returns a 1x1 pixel GIF, or forwards it to the token server with the referrer as a GET parameter for reporting
+
+var token_server = 'https://canarytokens.com';
+
+function handler(event) {
+    //console.log(event.request.uri);
```
Comment (jayjb) [inline aws-css-token-infra/CSSClonedSiteCFFunc/index.js:8]: Rm comment please

File: canarytokens/azure_css.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Returns: True on success, False otherwise
+    """
+    client = _auth_to_tenant(tenant_id)
+    if _check_if_custom_branding(client, tenant_id):
+        return (False, f"Installation failed: your tenant already has custom CSS, please manually add the following CSS to your branding:<br /><textarea>{css}</textarea>")
+    if not _install_custom_css(client, tenant_id, css):
+        # TODO: Better error handling
+        # Might as well remove ourselves anyways
+        _delete_self(client)
+        return (False, f"Installation failed: Unable to automaticall install the CSS, please manually add the following CSS to your branding:<br /><textarea>{css}</textarea>")
```
Comment (jayjb) [inline canarytokens/azure_css.py:53]: `automatically`

**Review by jayjb — APPROVED** [review]

**Pull request merged**